### DISCUSSION TITLE
What to do with "document", "window" vars when SSR is enabled?

### DISCUSSION THREAD

**Discussion by romsar** [inertiajs/inertia#1043] [Help]
So, now, after SSR configured and enabled, I must avoid all client stuff like window, document variables?
For example, in Laravel Jetstream with Inertia by default we have Modal.vue that contain this code:

```
document.body.style.overflow = 'hidden'
```

Because server does know document variable it will throw error: "ReferenceError: document is not defined".
That should I do in that situation to avoid this exception? Check if document is not undefined or that?

**Comment by ajnsn** [comment]
Hey @RomanSarvarov 

yep, at least Inertia itself is doing it for the `window` [this way](https://github.com/inertiajs/inertia/blob/master/packages/inertia/src/router.ts#L10).

**Reply by livevsonline** [reply]
Is this really and still the best solution? It seems a bit cumbersome to implement that check every place there is a document request?

**Reply by ajnsn** [reply]
Hey @livevsonline 

you should take a look at your JS framework docs how to deal with SSR, especially when it comes to the lifecycle (hooks). [Vue](https://vuejs.org/guide/scaling-up/ssr.html#component-lifecycle-hooks) does e.g. not call `mounted` or `updated`. Same applies to React (e.g. `componentDidMount`). 

If you have some cases where you need to render based on DOM objects then just render them in the correct lifecycle.

Hope this helps!

**Reply by livevsonline** [reply]
Well, I'm using Laravel, Jetstream and Inertia if I create a fresh project and implement the SSR it works perfectly when removing a `document.` from `Modal.vue`

However, that's not possible for the main website which has a few document calls, i could implement something like the suggested it just feels wrong doesn't it? idk.. 

Seems like [this](https://aaronfrancis.com/2021/vue-referenceerror-document-is-not-defined-at-insertstyleelement) person is having the exact same problem, he adds some code to `webpack.mix.js`, but that does not do the job for me, I still see the error

**Reply by ajnsn** [reply]
It looks like the Laravel guys do not have [any intention to support SSR at this time](https://github.com/laravel/jetstream/issues/987).

What I was trying to say is that in the end this has nothing to do with Inertia. You would run into this even when running SSR without Intertia. It's just Vue and SSR and yes, it could also be your bundler.

**Reply by ObjectAssign** [reply]
> Hey @RomanSarvarov
> 
> yep, at least Inertia itself is doing it for the `window` [this way](https://github.com/inertiajs/inertia/blob/master/packages/inertia/src/router.ts#L10).

Hey that link Is now broken, how did they solve this? also it looks like they are supporting it now from the link you submitted "any intention to support SSR at this time" ? :)

**Reply by ajnsn** [reply]
https://github.com/inertiajs/inertia/blob/79eb916a5832bed56aaf201ce07eb9f8b653526a/packages/core/src/router.ts#L35

```js
const isServer = typeof window === 'undefined'

if (! isServer) {
  // do something
}
```

**Reply by ObjectAssign** [reply]
> https://github.com/inertiajs/inertia/blob/79eb916a5832bed56aaf201ce07eb9f8b653526a/packages/core/src/router.ts#L35
> 
> ```js
> const isServer = typeof window === 'undefined'
> 
> if (! isServer) {
>   // do something
> }
> ```

I see, what is possible when a package uses the window object?

I use highcharts-vue which is a package from highcharts, its a bit outdated and needs options API instead of composition API..

It uses the window object, and when I try to run the inertia:start-ssr command it fails at: node_modules/highcharts-vue/dist/highcharts-vue.min.js

I have tried wrapping everything I can with that logic + using defineAsyncComponent to make it work, but it continues to fail

**Reply by reinink** [reply]
> I see, what is possible when a package uses the window object?

In these situations you're kind of out of luck from an SSR perspective. Most modern higher quality packages include SSR support, as SSR is very common these days with frameworks like Nuxt and Next.js.

If you really want to use the package I'd recommend conditionally loading it only on the client. You can do this with a package like [vue-client-only](https://www.npmjs.com/package/vue-client-only).
